compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -2570,9 +2571,25 @@ object Types {
               || adapted.info.eq(denot.info))
             adapted
           else this
-        result.setDenot(denot)
+        val lastDenot = result.lastDenotation
+        denot match
+          case denot: SymDenotation
+          if denot.validFor.firstPhaseId < ctx.phase.id
+            && lastDenot != null
+            && lastDenot.validFor.lastPhaseId > denot.validFor.firstPhaseId
+            && !lastDenot.isInstanceOf[SymDenotation] =>
+            // In this case the new SymDenotation might be valid for all phases, which means
+            // we would not recompute the denotation when travelling to an earlier phase, maybe
+            // in the next run. We fix that problem by creating a UniqueRefDenotation instead.
+            core.println(i"overwrite ${result.toString} / ${result.lastDenotation}, ${result.lastDenotation.getClass} with $denot at ${ctx.phaseId}")
+            result.setDenot(
+              UniqueRefDenotation(
```

Comment:
I think the comment in Compiler.scala about termrefs having only SymDenotation after erasure needs to be updated, but I'm not sure how to reword it to explain what we're doing here:
https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/Compiler.scala#L27-L33

Reply:
done